File: sdk/swift/Sources/TelecomSDK/SubscriberAPI.swift
import Foundation

/// API for subscriber management
public class SubscriberAPI {
    private let client: HTTPClient
    
    public init(client: HTTPClient) {
        self.client = client
    }
    
    /// Get subscriber by ID
    public func get(id: Int64) async throws -> Subscriber {
        return try await client.get(path: "/v1/subscribers/\(id)")
    }
    
    /// List subscribers with pagination
    public func list(page: Int = 1, pageSize: Int = 50, status: SubscriberStatus? = nil) async throws -> SubscriberList {
        var params: [String: String] = [
            "page": "\(page)",
            "page_size": "\(pageSize)"
        ]
        
        if let status = status {
            params["status"] = "\(status)"
        }
        
        return try await client.get(path: "/v1/subscribers", params: params)
    }
    
    /// Create a new subscriber
    public func create(_ request: CreateSubscriberRequest) async throws -> Subscriber {
        return try await client.post(path: "/v1/subscribers", body: request)
    }
    
    /// Update an existing subscriber
    public func update(id: Int64, request: UpdateSubscriberRequest) async throws -> Subscriber {
        return try await client.put(path: "/v1/subscribers/\(id)", body: request)
    }
    
    /// Delete a subscriber
    public func delete(id: Int64) async throws {
        try await client.delete(path: "/v1/subscribers/\(id)")
    }
    
    /// Suspend a subscriber
    public func suspend(id: Int64) async throws -> Subscriber {
        return try await client.post(path: "/v1/subscribers/\(id)/suspend", body: nil as EmptyRequest?)
    }
    
    /// Activate a suspended subscriber
    public func activate(id: Int64) async throws -> Subscriber {
        return try await client.post(path: "/v1/subscribers/\(id)/activate", body: nil as EmptyRequest?)
    }
}

private struct EmptyRequest: Encodable {}

File: sdk/kotlin/src/main/kotlin/com/telecom/SubscriberAPI.kt
package com.telecom

/**
 * API for subscriber management
 */
class SubscriberAPI(private val client: HTTPClient) {
    
    /**
     * Get subscriber by ID
     */
    suspend fun get(id: Long): Subscriber {
        return client.get("/v1/subscribers/$id")
    }
    
    /**
     * List subscribers with pagination
     */
    suspend fun list(page: Int = 1, pageSize: Int = 50, status: SubscriberStatus? = null): SubscriberList {
        val params = mutableMapOf(
            "page" to page.toString(),
            "page_size" to pageSize.toString()
        )
        
        status?.let { params["status"] = it.toString() }
        
        return client.get("/v1/subscribers", params)
    }
    
    /**
     * Create a new subscriber
     */
    suspend fun create(request: CreateSubscriberRequest): Subscriber {
        return client.post("/v1/subscribers", request)
    }
    
    /**
     * Update an existing subscriber
     */
    suspend fun update(id: Long, request: UpdateSubscriberRequest): Subscriber {
        return client.put("/v1/subscribers/$id", request)
    }
    
    /**
     * Delete a subscriber
     */
    suspend fun delete(id: Long) {
        client.delete("/v1/subscribers/$id")
    }
    
    /**
     * Suspend a subscriber
     */
    suspend fun suspend(id: Long): Subscriber {
        return client.post("/v1/subscribers/$id/suspend", null)
    }
    
    /**
     * Activate a suspended subscriber
     */
    suspend fun activate(id: Long): Subscriber {
        return client.post("/v1/subscribers/$id/activate", null)
    }
}
